extension/src/datum_utils.rs

```diff
@@ -198,55 +198,98 @@ flat_serialize_macro::flat_serialize! {
-//         S: serde::Serializer {
-//         let mut type_output = 0;
-//         let mut typ_is_varlena = false;
-//         let mut flinfo = unsafe {
-//             std::mem::MaybeUninit::zeroed().assume_init()
-//         };
-//         unsafe {
-//             pg_sys::getTypeOutputInfo(self.0.type_oid.0, &mut type_output, &mut typ_is_varlena);
-//             fmgr_info(type_output, &flinfo);
-//         }
-//         let count = self.0.into_iter().count();
-//         let mut seq = serializer.serialize_seq(Some(count))?;
-//         for element in self.0.into_iter() {
-//             let chars = pg_sys::OutputFunctionCall(&flinfo, element);
-//             let cstr = unsafe {
-//                 std::ffi::CStr::from_ptr(chars)
-//             };
-//             seq.serialize_element(cstr.to_str().unwrap())?;
-//         }
-//         seq.end()
-//     }
-// }
+impl<'a> Serialize for DatumStore<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer
+    {
+        let mut type_output = 0;
+        let mut typ_is_varlena = false;
+        let mut flinfo = unsafe {
+            std::mem::MaybeUninit::zeroed().assume_init()
+        };
+        unsafe {
+            pg_sys::getTypeOutputInfo(self.type_oid.0, &mut type_output, &mut typ_is_varlena);
+            pg_sys::fmgr_info(type_output, &mut flinfo);
+        }
+        let count = self.iter().count();
+        let mut seq = serializer.serialize_seq(Some(count + 1))?;
+        seq.serialize_element(&self.type_oid.0)?;
+        for element in self.iter() {
+            let chars = unsafe{pg_sys::OutputFunctionCall(&mut flinfo, element)};
+            let cstr = unsafe {
+                std::ffi::CStr::from_ptr(chars)
+            };
+            seq.serialize_element(cstr.to_str().unwrap())?;
+        }
+        seq.end()
+    }
+}
+
+impl<'a, 'de> Deserialize<'de> for DatumStore<'a> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>
+    {
+        struct DatumStoreVisitor<'a>(std::marker::PhantomData<core::cell::Cell<&'a ()>>);
+
+        impl<'de, 'a> Visitor<'de> for DatumStoreVisitor<'a> {
+            type Value = DatumStore<'a>;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("a sequence encoding a DatumStore object")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                // borrowed from serialization/types.rs
+                unsafe fn get_struct<T>(tuple: pg_sys::HeapTuple) -> *mut T {
+                    //((char *) ((TUP)->t_data) + (TUP)->t_data->t_hoff)
+                    let t_data: *mut u8 = (*tuple).t_data.cast();
+                    let t_hoff = (*(*tuple).t_data).t_hoff;
+                    t_data.add(t_hoff as usize).cast()
+                }
```

Comment:
I think we should only have one `fn get_struct()`, probably in `serialization`

Reply:
Agreed, I'll make the serialization one crate public.

Reply:
Actually this is no longer needed now that I'm not getting typmod, so I'm not bothering making get_struct available.